packages/core/src/components/dialog/multistepDialog.tsx

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2020 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import classNames from "classnames";
+import * as React from "react";
+import { polyfill } from "react-lifecycles-compat";
+
+import { AbstractPureComponent2, Classes, Utils } from "../../common";
+import { DISPLAYNAME_PREFIX } from "../../common/props";
+import { Button, IButtonProps } from "../button/buttons";
+import { Dialog, IDialogProps } from "./dialog";
+import { IStepProps, Step, StepId } from "./step";
+
+type StepElement = React.ReactElement<IStepProps & { children: React.ReactNode }>;
+
+export interface IMultistepDialogProps extends IDialogProps {
+    /**
+     * Whether to disable the next button in the panel footer.
+     */
+    disableNext?: boolean;
+
+    /**
+     * Props for the button to display on the final step.
+     */
+    finalButtonProps?: Partial<IButtonProps>;
+
+    /**
+     * A callback that is invoked when the user selects a different step by clicking on back, next, or a step itself.
+     */
+    onChange?(newStepId: StepId, prevStepId: StepId | undefined, event: React.MouseEvent<HTMLElement>): void;
+}
+
+interface IMultistepDialogState {
+    lastViewedIndex: number;
+    selectedIndex: number;
+}
+
+const PADDING_BOTTOM = 0;
+
+const MIN_WIDTH = 800;
+
+@polyfill
+export class MultistepDialog extends AbstractPureComponent2<IMultistepDialogProps, IMultistepDialogState> {
+    public static displayName = `${DISPLAYNAME_PREFIX}.MultistepDialog`;
+
+    public static defaultProps: Partial<IMultistepDialogProps> = {
+        canOutsideClickClose: true,
+        isOpen: false,
+    };
+
+    public constructor(props: IMultistepDialogProps) {
+        super(props);
+        this.state = this.getDefaultState();
+    }
+
+    public render() {
+        return (
+            <Dialog {...this.props} style={this.getDialogStyle()}>
+                <div className={Classes.MULTISTEP_DIALOG_PANELS}>
+                    {this.renderLeftPanel()}
+                    {this.maybeRenderRightPanel()}
+                </div>
+            </Dialog>
+        );
+    }
+
+    public componentDidUpdate(prevProps: IMultistepDialogProps) {
+        if (!prevProps.isOpen && this.props.isOpen) {
+            this.setState(this.getDefaultState());
+        }
+    }
+
+    private getDefaultState() {
```

Comment:
If we only need to define `lastViewedIndex` and `selectedIndex`, probably a little clearer to just define those in an `INITIAL_STATE` const and then set state to `INITIAL_STATE` in the constructor and `componentDidUpdate`